### PR TITLE
ACCUMULO-4612 - Simplify Accumulo memory configuration

### PR DESCRIPTION
**Pull request by mikewalch**
* Tablet server memory setttings now default to a percentage
  of Java heap space given to process.  Default settings will
  work for various heap space settings and fixed memory
  settings are still supported.
* Remove 'accumulo create-config' command as it no longer needed.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 1]
```diff
@@ -253,13 +252,13 @@
...
       + "size is ok because of group commit."),
-  TSERV_TOTAL_MUTATION_QUEUE_MAX("tserver.total.mutation.queue.max", "50M", PropertyType.MEMORY,
+  TSERV_TOTAL_MUTATION_QUEUE_MAX("tserver.total.mutation.queue.max", "5%", PropertyType.MEMORY,
       "The amount of memory used to store write-ahead-log mutations before flushing them."),
   TSERV_TABLET_SPLIT_FINDMIDPOINT_MAXOPEN("tserver.tablet.split.midpoint.files.max", "300", PropertyType.COUNT,
       "To find a tablets split points, all index files are opened. This setting determines how many index "
           + "files can be opened at once. When there are more index files than this setting multiple passes "
           + "must be made, which is slower. However opening too many files at once can cause problems."),
-  TSERV_WALOG_MAX_SIZE("tserver.walog.max.size", "1G", PropertyType.MEMORY,
+  TSERV_WALOG_MAX_SIZE("tserver.walog.max.size", "33%", PropertyType.MEMORY,
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: This is a file size limit.   Should probably be left at 1G

File: docs/src/main/asciidoc/chapters/administration.txt [hunk 2]
```diff
@@ -404,6 +399,51 @@ and BatchScanner passing in the name of the context, app1 in the example above.
 iterators to load classes from the locations defined by the context. Passing the context name to the scanners allows you to override the table setting
 to load only scan time iterators from a different location. 
 
+=== Vendor-specific configuration
```
Comment (keith-turner) [inline docs/src/main/asciidoc/chapters/administration.txt]: I think this info would better on the website, where it can be easily updated and appended to.

We could create the web page now w/o any links to it on the website .. and a disclaimer about the info being for 2.0.0-SNAP.  Then 2.0.0-SNAP docs could point to it.

File: INSTALL.md [hunk 3]
```diff
@@ -36,62 +36,49 @@ to manage Accumulo:
...
- * **Location of Zookeeper and Hadoop jars** :  Setting `ZOOKEEPER_HOME` and
-   `HADOOP_PREFIX` in `accumulo-env.sh` will help Accumulo find these jars
-   when using the default setting for `general.classpaths` in accumulo-site.xml.
-
-If Accumulo has problems later on finding jars, then run `bin/accumulo
-classpath` to print out info about where Accumulo is finding jars.  If the
-settings mentioned above are correct, then inspect `general.classpaths` in
-`accumulo-site.xml`.
+3. (Optional) The Accumulo master is configured by default to use 512MB while the garbage collector
+   and monitor is configure to use 256MB. These settings can be changed in the `JAVA_OPTS` section
```
Comment (keith-turner) [inline INSTALL.md]: `monitor is configure to use`

File: INSTALL.md [hunk 4]
```diff
@@ -36,62 +36,49 @@ to manage Accumulo:
...
-  `instance.volumes` to `hdfs://192.168.1.9:9000/accumulo`.
- * **Location of Zookeeper and Hadoop jars** :  Setting `ZOOKEEPER_HOME` and
-   `HADOOP_PREFIX` in `accumulo-env.sh` will help Accumulo find these jars
-   when using the default setting for `general.classpaths` in accumulo-site.xml.
-
-If Accumulo has problems later on finding jars, then run `bin/accumulo
-classpath` to print out info about where Accumulo is finding jars.  If the
-settings mentioned above are correct, then inspect `general.classpaths` in
-`accumulo-site.xml`.
+3. (Optional) The Accumulo master is configured by default to use 512MB while the garbage collector
```
Comment (keith-turner) [inline INSTALL.md]: Can you word this differently so that the exact memory amounts are not specified here?  If those defaults are changed in the env file, then the person making the change would need to know to make the changes here.  I think its better to not duplicate the info here.

File: assemble/conf/accumulo-env.sh [hunk 5]
```diff
@@ -36,6 +36,22 @@ export HADOOP_CONF_DIR="${HADOOP_CONF_DIR:-${HADOOP_PREFIX}/etc/hadoop}"
 ## Zookeeper installation
 export ZOOKEEPER_HOME="${ZOOKEEPER_HOME:-/path/to/zookeeper}"
 
+##########################
+# Build CLASSPATH variable
+##########################
+
+## Adds external Hadoop & Zookeeper dependencies to CLASSPATH. See "Vendor configuration" section of Accumulo user manual
+## for different settings if you installed vendor's distribution of Hadoop or Zookeeper.
```
Comment (keith-turner) [inline assemble/conf/accumulo-env.sh]: If we put info about vendor specific configs on website, then we could put a link to that info here.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 6]
```diff
@@ -170,10 +170,9 @@
   // general properties
   GENERAL_PREFIX("general.", null, PropertyType.PREFIX,
       "Properties in this category affect the behavior of accumulo overall, but do not have to be consistent throughout a cloud."),
-  GENERAL_CLASSPATHS(AccumuloClassLoader.CLASSPATH_PROPERTY_NAME, AccumuloClassLoader.ACCUMULO_CLASSPATH_VALUE, PropertyType.STRING,
+  GENERAL_CLASSPATHS(AccumuloClassLoader.CLASSPATH_PROPERTY_NAME, "", PropertyType.STRING,
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Should we deprecate this property?

File: docs/src/main/asciidoc/chapters/administration.txt [hunk 7]
```diff
@@ -125,6 +124,12 @@ and specify the following:
 . Enter the location of ZooKeeper for +$ZOOKEEPER_HOME+
 . Optionally, choose a different location for Accumulo logs using +$ACCUMULO_LOG_DIR+
 
+Accumulo uses +HADOOP_PREFIX+ and +ZOOKEEPER_HOME+ to locate Hadoop and Zookeeper jars
+and add them the +CLASSPATH+ variable. If you are running a vendor-specific release of Hadoop
+or Zookeeper, see the `Vendor configuration` section as you may need to change how your `CLASSPATH`
```
Comment (keith-turner) [inline docs/src/main/asciidoc/chapters/administration.txt]: if info is on website, then could link to it from here

File: core/src/main/java/org/apache/accumulo/core/conf/PropertyType.java [hunk 8]
```diff
@@ -45,8 +45,8 @@
           + "Examples of invalid durations are '1w', '1h30m', '1s 200ms', 'ms', '', and 'a'.\n"
           + "Unless otherwise stated, the max value for the duration represented in milliseconds is " + Long.MAX_VALUE),
 
-  MEMORY("memory", boundedUnits(0, Long.MAX_VALUE, false, "", "B", "K", "M", "G"),
-      "A positive integer optionally followed by a unit of memory (whitespace disallowed), as in 2G.\n"
+  MEMORY("memory", boundedUnits(0, Long.MAX_VALUE, false, "", "B", "K", "M", "G", "%"),
+      "A positive integer optionally followed by a unit of memory or percentage (whitespace disallowed), as in 2G.\n"
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/PropertyType.java]: Need to describe behavior of percentage.

File: core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java [hunk 9]
```diff
@@ -209,13 +209,19 @@ static public long getMemoryInBytes(String str) {
...
+          }
+          return Runtime.getRuntime().maxMemory() * percent / 100;
         default:
           return Long.parseLong(str);
       }
       return Long.parseLong(str.substring(0, str.length() - 1)) << multiplier;
     } catch (Exception ex) {
       throw new IllegalArgumentException("The value '" + str + "' is not a valid memory setting. A valid value would a number "
-          + "possibily followed by an optional 'G', 'M', 'K', or 'B'.");
+          + "possibly followed by an optional 'G', 'M', 'K', or 'B'.");
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: Since this is catching exception, it possible that something could go wrong parsing a `%`.  So error message should include `%` as valid suffix.

File: docs/src/main/asciidoc/chapters/administration.txt [hunk 10]
```diff
@@ -364,17 +357,19 @@ consideration. There is no enforcement of these warnings via the API.
 
 ==== Configuring the ClassLoader
 
-Accumulo loads classes from the locations specified in the +general.classpaths+ property. Additionally, Accumulo will load classes
-from the locations specified in the +general.dynamic.classpaths+ property and will monitor and reload them if they change. The reloading 
-feature is useful during the development and testing of iterators as new or modified iterator classes can be deployed to Accumulo without
-having to restart the database.
+Accumulo builds its Java classpath in +accumulo-env.sh+.  After Accumulo application have started, they will loads classes from the locations
```
Comment (keith-turner) [inline docs/src/main/asciidoc/chapters/administration.txt]: `they will loads classes`

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java (hunk 6 above)
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: I think it should be deprecated. I think users should avoid setting up the classpath in accumulo-site.xml.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java (hunk 1 above)
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Ok.  I will change.

File: core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java [hunk 11]
```diff
@@ -209,13 +209,19 @@ static public long getMemoryInBytes(String str) {
         case 'B':
           multiplier = 0;
           break;
+        case '%':
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: I'm wondering if it would make more sense to create a new `PropertyType` instead of piggy-backing on `MEMORY`. Is it confusing to allow percent-values for some of these other properties which we really never expect users to do?

For example, some properties like `table.file.blocksize` or `tserver.compaction.major.throughput` make no sense to allow a percentage of total heap.

File: docs/src/main/asciidoc/chapters/administration.txt (hunk 2 above)
Comment (mikewalch) [inline docs/src/main/asciidoc/chapters/administration.txt]: I would prefer if it stayed in the user manual as these settings could vary between minor versions of Accumulo. If vendors release new versions and these docs need to be updated,  the user manual can be updated in the Accumulo repo and pushed to the website without creating a new bugfix release.

File: INSTALL.md (hunk 4 above)
Comment (mikewalch) [inline INSTALL.md]: Should be fixed.

File: core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java (hunk 11 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: Could have a new type `BYTES` which has everything but the `%`.  And `MEMORY` could be what `BYTES` supports plus `%`.

Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: I would rather not create another `PropertyType`. If the defaults are fixed memory, users probably won't switch to percentage.  If you are worried about certain properties, I could add some text to the description of those properties that fixed memory amounts should be used.  Even if we limit percentage, users can still misconfigure things by setting fixed memory values that are too high or low for that property.

File: docs/src/main/asciidoc/chapters/administration.txt (hunk 2 above)
Comment (keith-turner) [inline docs/src/main/asciidoc/chapters/administration.txt]: > If vendors release new versions and these docs need to be updated, the user manual can be updated in the Accumulo repo and pushed to the website without creating a new bugfix release.

That addresses part of my concern.  Still does not allow linking to the documentation from accumulo-env.sh.  However, the inability to link between different kinds of Accumulo documentation is a more general issue that should not be addresses here.

File: core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java (hunk 11 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: > If you are worried about certain properties, I could add some text to the description of those properties that fixed memory amounts should be used

That would also assuage my fears. I am not worried about the general misconfiguration -- just users trying to set percentages where it makes no sense.

**Comment by joshelser** [comment]
With the percentages essentially being transformed into discrete values, our same sanity checks should be in place that would prevent users from providing configuration that exceeds the heap size?

I think it's just the one: `'-Xmx' >= (block caches) + (non-native-maps)`? Do we need to update that check? Does anyone remember where that check is? :)

**Comment by keith-turner** [comment]
> Does anyone remember where that check is? 

I think the check is in `TabletServerResourceManager` constructor.  I think the check will be ok as it uses the resolved memory values, but not completely sure.

File: core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java (hunk 11 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: > Even if we limit percentage, users can still misconfigure things by setting fixed memory values that are too high or low for that property.

Thats true.  Using types just narrows the set of possible mistakes, it does not eliminate mistakes.

**Comment by joshelser** [comment]
> I think the check is in TabletServerResourceManager constructor. I think the check will be ok as it uses the resolved memory values, but not completely sure.

Your memory is much better than mine.

https://github.com/apache/accumulo/blob/94cdcc4d3f0a8ccf95894f206cb71e6117f4e51d/server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServerResourceManager.java#L192-L202

I would agree that I think this will be fine.

File: core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java (hunk 11 above)
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: I pushed another commit that creates a `BYTES` type.

File: docs/src/main/asciidoc/chapters/administration.txt (hunk 2 above)
Comment (ctubbsii) [inline docs/src/main/asciidoc/chapters/administration.txt]: The user manual is tied to a version of Accumulo. It cannot be updated over time between releases. The copy on the website can, but the "1.8.0" user manual is fixed at a point in time and is distributed with the tarball.

I agree with Keith it's better to put this on the website, outside the user manual. The user manual should have generic information applicable to everybody to keep it succinct. Specific documentation for a particular environment can be discussed in a blog, which refers to the appropriate section of the user manual.

In other words:

usermanual: "Section X: You should configure your CLASSPATH by doing Y"
blog: "You can set your class path for HDP to ..., according to section X" (or similar)

This also makes more sense from a versioning perspective. The classpath for HDP might not change, but the instructions in "Section X" (in my example) might be different. Keeping them separate like this keeps them both relevant longer.

File: core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java [hunk 12]
```diff
@@ -165,30 +165,35 @@ private void checkType(Property property, PropertyType type) {
...
+   * Interprets a string specifying bytes. A bytes type is specified as a long integer followed by an optional B (bytes), K (KB), M (MB), or G (GB).
    *
    * @param str
-   *          string value
-   * @return interpreted memory size
+   *          String value
+   * @return interpreted memory size in bytes
    */
-  static public long getMemoryInBytes(String str) {
+  static public long getBytes(String str) {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java]: getBytes is too common of a method name (Example `String#getBytes()`). The type is still for bytes of memory, so there's no reason to change that. Keeping the old name makes the changeset much smaller. The new one could be something more like: `getMemoryRatioAsBytes()`, and could be parsed as either a ratio or fall back to bytes (if not ending with '%').

File: assemble/conf/accumulo-site.xml [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  limitations under the License.
+-->
+<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
+<!-- This is the main configuration file for Apache Accumulo. Available configuration properties
+  (and their default values) can be found in the user manual (docs/accumulo_user_manual.html). -->
+<configuration>
+  <!-- Set location in HDFS where Accumulo will store data -->
+  <property>
+    <name>instance.volumes</name>
+    <value>hdfs://localhost:9000/accumulo</value>
```
Comment (joshelser) [inline assemble/conf/accumulo-site.xml]: 8020

Comment (ctubbsii) [inline assemble/conf/accumulo-site.xml]: Thanks @joshelser :)

File: assemble/bin/accumulo [hunk 14]
```diff
@@ -80,9 +65,6 @@ function main() {
     JAVA=($ACCUMULO_JAVA_PREFIX $JAVA)
```
Comment (mikewalch) [inline assemble/bin/accumulo]: I am ok with splitting on whitespace.  i tested this current code with numactl commands and it worked fine.

File: docs/src/main/asciidoc/chapters/administration.txt (hunk 2 above)
Comment (mikewalch) [inline docs/src/main/asciidoc/chapters/administration.txt]: I pushed another commit that removes the vendor-specific documentation from the user manual.

**Comment by mikewalch** [comment]
I would like to merge today if there are no objections.

File: INSTALL.md [hunk 15]
```diff
@@ -36,62 +36,49 @@ to manage Accumulo:
...
 
-Accumulo is configured by the files `accumulo-site.xml` and `accumulo-env.sh` in the `conf/`
-directory. You can either edit these files for your environment or run the command below which will
-overwrite them with files configured for your environment.
+1. Run `accumulo-util build-native` to build native code.  If this command fails, disable
+   native maps by setting `tserver.memory.maps.native.enabled` to `false`.
 
-    accumulo-util create-config
+2. Set `instance.volumes` to HDFS location where Accumulo will store data. If your namenode
+   is running at 192.168.1.9:9000 and you want to store data in `/accumulo` in HDFS, then set
```
Comment (joshelser) [inline INSTALL.md]: If you did the s/9000/8020/ elsewhere, could also do it here

Comment (mikewalch) [inline INSTALL.md]: This is now fixed.

File: core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java [hunk 16]
```diff
@@ -49,12 +50,12 @@ public void testGetTimeInMillis() {
 
   @Test(expected = IllegalArgumentException.class)
   public void testGetMemoryInBytesFailureCases1() throws Exception {
-    AccumuloConfiguration.getMemoryInBytes("42x");
+    AccumuloConfiguration.getFixedMemoryAsBytes("42x");
```
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java]: Could duplicate these two failure test for getMemoryAsBytes

File: core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java [hunk 17]
```diff
@@ -25,16 +25,17 @@
...
-    assertEquals(42l, AccumuloConfiguration.getMemoryInBytes("42b"));
-    assertEquals(42l, AccumuloConfiguration.getMemoryInBytes("42B"));
-    assertEquals(42l * 1024l, AccumuloConfiguration.getMemoryInBytes("42K"));
-    assertEquals(42l * 1024l, AccumuloConfiguration.getMemoryInBytes("42k"));
-    assertEquals(42l * 1024l * 1024l, AccumuloConfiguration.getMemoryInBytes("42M"));
-    assertEquals(42l * 1024l * 1024l, AccumuloConfiguration.getMemoryInBytes("42m"));
-    assertEquals(42l * 1024l * 1024l * 1024l, AccumuloConfiguration.getMemoryInBytes("42G"));
-    assertEquals(42l * 1024l * 1024l * 1024l, AccumuloConfiguration.getMemoryInBytes("42g"));
-
+    assertEquals(42l, AccumuloConfiguration.getFixedMemoryAsBytes("42"));
```
Comment (mikewalch) [inline core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java]: This was added.

File: core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java (hunk 16 above)
Comment (mikewalch) [inline core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java]: This was added.
